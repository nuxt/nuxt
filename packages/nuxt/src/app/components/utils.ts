import { h } from 'vue'
import type { Component, RendererNode } from 'vue'
// eslint-disable-next-line
import { isString, isPromise, isArray, isObject } from '@vue/shared'
import destr from 'destr'

/**
 * Internal utility
 *
 * @private
 */
export const _wrapIf = (component: Component, props: any, slots: any) => {
  props = props === true ? {} : props
  return { default: () => props ? h(component, props, slots) : slots.default?.() }
}

// eslint-disable-next-line no-use-before-define
export type SSRBuffer = SSRBufferItem[] & { hasAsync?: boolean }
export type SSRBufferItem = string | SSRBuffer | Promise<SSRBuffer>

/**
 * create buffer retrieved from @vue/server-renderer
 *
 * @see https://github.com/vuejs/core/blob/9617dd4b2abc07a5dc40de6e5b759e851b4d0da1/packages/server-renderer/src/render.ts#L57
 * @private
 */
export function createBuffer () {
  let appendable = false
  const buffer: SSRBuffer = []
  return {
    getBuffer (): SSRBuffer {
      return buffer
    },
    push (item: SSRBufferItem) {
      const isStringItem = isString(item)
      if (appendable && isStringItem) {
        buffer[buffer.length - 1] += item as string
      } else {
        buffer.push(item)
      }
      appendable = isStringItem
      if (isPromise(item) || (isArray(item) && item.hasAsync)) {
        buffer.hasAsync = true
      }
    }
  }
}

const TRANSLATE_RE = /&(nbsp|amp|quot|lt|gt);/g
const NUMSTR_RE = /&#(\d+);/gi
export function decodeHtmlEntities (html: string) {
  const translateDict = {
    nbsp: ' ',
    amp: '&',
    quot: '"',
    lt: '<',
    gt: '>'
  } as const
  return html.replace(TRANSLATE_RE, function (_, entity: keyof typeof translateDict) {
    return translateDict[entity]
  }).replace(NUMSTR_RE, function (_, numStr: string) {
    const num = parseInt(numStr, 10)
    return String.fromCharCode(num)
  })
}

/**
 * helper for NuxtIsland to generate a correct array for scoped data
 */
export function vforToArray (source: any): any[] {
  if (isArray(source)) {
    return source
  } else if (isString(source)) {
    return source.split('')
  } else if (typeof source === 'number') {
    if (process.dev && !Number.isInteger(source)) {
      console.warn(`The v-for range expect an integer value but got ${source}.`)
    }
    const array = []
    for (let i = 0; i < source; i++) {
      array[i] = i
    }
    return array
  } else if (isObject(source)) {
    if (source[Symbol.iterator as any]) {
      return Array.from(source as Iterable<any>, item =>
        item
      )
    } else {
      const keys = Object.keys(source)
      const array = new Array(keys.length)
      for (let i = 0, l = keys.length; i < l; i++) {
        const key = keys[i]
        array[i] = source[key]
      }
      return array
    }
  }
  return []
}

export function getFragmentHTML (element: RendererNode | null) {
  if (element) {
    if (element.nodeName === '#comment' && element.nodeValue === '[') {
      return getFragmentChildren(element)
    }
    return [element.outerHTML]
  }
  return []
}

function getFragmentChildren (element: RendererNode | null, blocks: string[] = []) {
  if (element && element.nodeName) {
    if (isEndFragment(element)) {
      return blocks
    } else if (!isStartFragment(element)) {
      blocks.push(element.outerHTML)
    }

    getFragmentChildren(element.nextSibling, blocks)
  }
  return blocks
}

function isStartFragment (element: RendererNode) {
  return element.nodeName === '#comment' && element.nodeValue === '['
}

function isEndFragment (element: RendererNode) {
  return element.nodeName === '#comment' && element.nodeValue === ']'
}
const SLOT_PROPS_RE = /<div[^>]*nuxt-ssr-slot-name="([^"]*)" nuxt-ssr-slot-data="([^"]*)"[^/|>]*>/g

export function getSlotProps (html: string) {
  const slotsDivs = html.matchAll(SLOT_PROPS_RE)
  const data: Record<string, any> = {}
  for (const slot of slotsDivs) {
    const [_, slotName, json] = slot
    const slotData = destr(decodeHtmlEntities(json))
    data[slotName] = slotData
  }
  return data
}
