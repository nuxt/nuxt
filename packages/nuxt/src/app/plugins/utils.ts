const VALID_ISLAND_KEY_RE = /^[a-z][a-z\d-]*_[a-z\d]+$/i
/* @__PURE__ */
export function isValidIslandKey (key: string): boolean {
  return typeof key === 'string' && VALID_ISLAND_KEY_RE.test(key) && key.length <= 100
}
