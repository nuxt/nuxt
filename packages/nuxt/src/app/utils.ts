export function toArray<T> (value: T | T[] | undefined): T[] {
  return Array.isArray(value) ? value : (value ? [value] : [])
}
