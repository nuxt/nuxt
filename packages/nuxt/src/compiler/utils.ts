import { type ParsedStaticImport, findStaticImports, parseStaticImport } from 'mlly'
import { parseAndWalk, walk } from 'oxc-walker'
import type { ScanPlugin } from '@nuxt/schema'
import type { ParseResult } from 'oxc-parser'

/**
 * Creates a context object for scan plugins scoped to a specific file being scanned.
 * All plugins scanning this file will share the same context instance.
 */
export function createScanPluginContext (code: string, filePath: string) {
  // create the context for sharing the parse result between plugins
  let parseResult: ParseResult | null = null
  let parsedStaticImports: ParsedStaticImport[] | null = null

  const pluginScanThisContext: ThisParameterType<ScanPlugin['scan']> = {
    walkParsed: (...args) => {
      if (parseResult) {
        const options: Parameters<typeof walk>[1] = typeof args[0] === 'function' ? { enter: args[0] } : args[0]
        walk(parseResult.program, options)
        return parseResult
      }

      parseResult = parseAndWalk.call(null, code, filePath, args[0])
      return parseResult
    },
    getParsedStaticImports: () => {
      if (parsedStaticImports) {
        return parsedStaticImports
      }

      const imports = findStaticImports(code)
      parsedStaticImports = imports.map(i => parseStaticImport(i))
      return parsedStaticImports
    },
  }
  return pluginScanThisContext
}

export function matchWithStringOrRegex (value: string, matcher: string | RegExp) {
  if (typeof matcher === 'string') {
    return value === matcher
  } else if (matcher instanceof RegExp) {
    return matcher.test(value)
  }

  return false
}
