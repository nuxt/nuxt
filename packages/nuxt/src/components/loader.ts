import { pathToFileURL } from 'node:url'
import { createUnplugin } from 'unplugin'
import { parseQuery, parseURL } from 'ufo'
import type { Component, ComponentsOptions } from '@nuxt/schema'
import { genDynamicImport, genImport } from 'knitwork'
import MagicString from 'magic-string'
import { pascalCase } from 'scule'
import { resolve } from 'pathe'
import { distDir } from '../dirs'

interface LoaderOptions {
  getComponents (): Component[]
  mode: 'server' | 'client'
  sourcemap?: boolean
  transform?: ComponentsOptions['transform']
  experimentalComponentIslands?: boolean
}

function isVueTemplate (id: string) {
  // Bare `.vue` file (in Vite)
  if (id.endsWith('.vue')) {
    return true
  }

  const { search } = parseURL(decodeURIComponent(pathToFileURL(id).href))
  if (!search) {
    return false
  }

  const query = parseQuery(search)

  // Macro
  if (query.macro) {
    return true
  }

  // Non-Vue or Styles
  if (!('vue' in query) || query.type === 'style') {
    return false
  }

  // Query `?vue&type=template` (in webpack or external template)
  return true
}

export const loaderPlugin = createUnplugin((options: LoaderOptions) => {
  const exclude = options.transform?.exclude || []
  const include = options.transform?.include || []
  const serverComponentRuntime = resolve(distDir, 'components/runtime/server-component')

  return {
    name: 'nuxt:components-loader',
    enforce: 'post',
    transformInclude (id) {
      if (exclude.some(pattern => id.match(pattern))) {
        return false
      }
      if (include.some(pattern => id.match(pattern))) {
        return true
      }
      return isVueTemplate(id)
    },
    transform (code, id) {
      const components = options.getComponents()

      let num = 0
      const imports = new Set<string>()
      const map = new Map<Component, string>()
      const s = new MagicString(code)

      // replace `_resolveComponent("...")` to direct import
      s.replace(/(?<=[ (])_?resolveComponent\(\s*["'](lazy-|Lazy)?([^'"]*?)["'][\s,]*[^)]*\)/g, (full: string, lazy: string, name: string) => {
        const component = findComponent(components, name, options.mode)
        if (component) {
          let identifier = map.get(component) || `__nuxt_component_${num++}`
          map.set(component, identifier)

          const isServerOnly = component.mode === 'server' &&
            !components.some(c => c.pascalName === component.pascalName && c.mode === 'client')
          if (isServerOnly) {
            imports.add(genImport(serverComponentRuntime, [{ name: 'createServerComponent' }]))
            imports.add(`const ${identifier} = createServerComponent(${JSON.stringify(name)})`)
            if (!options.experimentalComponentIslands) {
              console.warn(`Standalone server components (\`${name}\`) are not yet supported without enabling \`experimental.componentIslands\`.`)
            }
            return identifier
          }

          const isClientOnly = component.mode === 'client'
          if (isClientOnly) {
            imports.add(genImport('#app/components/client-only', [{ name: 'createClientOnly' }]))
            identifier += '_client'
          }

          if (lazy) {
            imports.add(genImport('vue', [{ name: 'defineAsyncComponent', as: '__defineAsyncComponent' }]))
            identifier += '_lazy'
            imports.add(`const ${identifier} = /*#__PURE__*/ __defineAsyncComponent(${genDynamicImport(component.filePath, { interopDefault: true })}${isClientOnly ? '.then(c => createClientOnly(c))' : ''})`)
          } else {
            imports.add(genImport(component.filePath, [{ name: component.export, as: identifier }]))

            if (isClientOnly) {
              imports.add(`const ${identifier}_wrapped = /*#__PURE__*/ createClientOnly(${identifier})`)
              identifier += '_wrapped'
            }
          }

          return identifier
        }
        // no matched
        return full
      })

      if (imports.size) {
        s.prepend([...imports, ''].join('\n'))
      }

      if (s.hasChanged()) {
        return {
          code: s.toString(),
          map: options.sourcemap
            ? s.generateMap({ source: id, includeContent: true })
            : undefined
        }
      }
    }
  }
})

function findComponent (components: Component[], name: string, mode: LoaderOptions['mode']) {
  const id = pascalCase(name).replace(/["']/g, '')
  // Prefer exact match
  const component = components.find(component => id === component.pascalName && ['all', mode, undefined].includes(component.mode))
  if (component) { return component }

  const otherModeComponent = components.find(component => id === component.pascalName)

  // Render client-only components on the server with <ServerPlaceholder> (a simple div)
  if (mode === 'server' && otherModeComponent) {
    return components.find(c => c.pascalName === 'ServerPlaceholder')
  }

  // Return the other-mode component in all other cases - we'll handle createClientOnly
  // and createServerComponent above
  return otherModeComponent
}
