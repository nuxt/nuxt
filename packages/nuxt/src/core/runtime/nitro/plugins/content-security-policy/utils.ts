import { existsSync } from 'node:fs'
import { readFile, readdir } from 'node:fs/promises'
import { isAbsolute, join } from 'pathe'
import { createDefu } from 'defu'
import type { Nitro } from 'nitropack/types'
import type { ContentSecurityPolicyValue } from './types'

export function headerStringFromObject (optionValue: ContentSecurityPolicyValue | false) {
  // False value translates into empty header
  if (optionValue === false) {
    return ''
  }
  // Stringify the options passed as a JS object
  return Object.entries(optionValue)
    .filter(([, value]) => value !== false)
    .map(([directive, sources]) => {
      if (directive === 'upgrade-insecure-requests') {
        return 'upgrade-insecure-requests;'
      } else {
        const stringifiedSources = (typeof sources === 'string')
          ? sources
          : (sources as string[])
              .map(source => source.trim())
              .join(' ')
        return `${directive} ${stringifiedSources};`
      }
    })
    .join(' ')
}

export const defuReplaceArray = createDefu((obj, key, value) => {
  if (Array.isArray(obj[key]) || Array.isArray(value)) {
    obj[key] = value
    return true
  }
})

export function generateRandomNonce () {
  const array = new Uint8Array(18)
  crypto.getRandomValues(array)
  const nonce = btoa(String.fromCharCode(...array))
  return nonce
}

export async function generateHash (content: Buffer | string, hashAlgorithm: 'SHA-256' | 'SHA-384' | 'SHA-512') {
  let buffer: Uint8Array
  if (typeof content === 'string') {
    buffer = new TextEncoder().encode(content)
  } else {
    buffer = new Uint8Array(content)
  }
  const hashBuffer = await crypto.subtle.digest(hashAlgorithm, buffer)
  const base64 = btoa(String.fromCharCode(...new Uint8Array(hashBuffer)))
  const prefix = hashAlgorithm.replace('-', '').toLowerCase()
  return `${prefix}-${base64}`
}

export async function hashBundledAssets (nitro: Nitro) {
  const hashAlgorithm = 'SHA-384'
  const sriHashes: Record<string, string> = {}

  // Will be later necessary to construct url
  const { cdnURL: appCdnUrl = '', baseURL: appBaseUrl } = nitro.options.runtimeConfig.app

  // Go through all public assets folder by folder
  const publicAssets = nitro.options.publicAssets
  for (const publicAsset of publicAssets) {
    const { dir, baseURL = '' } = publicAsset

    if (existsSync(dir)) {
      // Node 16 compatibility maintained
      // Node 18.17+ supports recursive option on readdir
      // const entries = await readdir(dir, { withFileTypes: true, recursive: true })
      const entries = await readdir(dir, { withFileTypes: true })
      for (const entry of entries) {
        if (entry.isFile()) {
          // Node 16 compatibility maintained
          // Node 18.17+ supports entry.path on DirEnt
          // const fullPath = join(entry.path, entry.name)
          const path = join(dir, entry.name)
          const content = await readFile(path)
          const hash = await generateHash(content, hashAlgorithm)
          // construct the url as it will appear in the head template
          const fullPath = join(baseURL, entry.name)
          let url: string
          if (appCdnUrl) {
            // If the cdnURL option was set, the url will be in the form https://...
            const relativePath = isAbsolute(fullPath) ? fullPath.slice(1) : fullPath
            const abdsoluteCdnUrl = appCdnUrl.endsWith('/') ? appCdnUrl : appCdnUrl + '/'
            url = new URL(relativePath, abdsoluteCdnUrl).href
          } else {
            // If not, the url will be in a relative form: /_nuxt/...
            url = join('/', appBaseUrl, fullPath)
          }
          sriHashes[url] = hash
        }
      }
    }
  }

  return sriHashes
}
