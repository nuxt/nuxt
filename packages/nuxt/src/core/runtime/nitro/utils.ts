import { type H3Event, setResponseHeaders } from 'h3'
import type { NitroErrorHandler } from 'nitro/types'

export function defineNitroErrorHandler (
  handler: NitroErrorHandler,
): NitroErrorHandler {
  return handler
}

export function setSecurityHeaders (event: H3Event, allowjs = false) {
  setResponseHeaders(event, {
    // Prevent browser from guessing the MIME types of resources.
    'X-Content-Type-Options': 'nosniff',
    // Prevent error page from being embedded in an iframe
    'X-Frame-Options': 'DENY',
    // Prevent browsers from sending the Referer header
    'Referrer-Policy': 'no-referrer',
    // Disable the execution of any js
    'Content-Security-Policy': allowjs
      ? 'script-src \'self\' \'unsafe-inline\'; object-src \'none\'; base-uri \'self\';'
      : 'script-src \'none\'; frame-ancestors \'none\';',
  })
}
