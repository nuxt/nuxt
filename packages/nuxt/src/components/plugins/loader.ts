import { createUnplugin } from 'unplugin'
import { genDynamicImport, genImport } from 'knitwork'
import MagicString from 'magic-string'
import { pascalCase } from 'scule'
import { relative } from 'pathe'
import type { Component, ComponentsOptions } from 'nuxt/schema'

import { tryUseNuxt } from '@nuxt/kit'
import { QUOTE_RE, SX_RE, isVue } from '../../core/utils'
import { installNuxtModule } from '../../core/features'
import { logger } from '../../utils'

interface LoaderOptions {
  getComponents (): Component[]
  mode: 'server' | 'client'
  srcDir: string
  serverComponentRuntime: string
  clientDelayedComponentRuntime: string
  sourcemap?: boolean
  transform?: ComponentsOptions['transform']
  experimentalComponentIslands?: boolean
}

const REPLACE_COMPONENT_TO_DIRECT_IMPORT_RE = /(?<=[ (])_?resolveComponent\(\s*(?<quote>["'`])(?<lazy>lazy-|Lazy(?=[A-Z]))?(?<modifier>Idle|Visible|idle-|visible-|Interaction|interaction-|MediaQuery|media-query-|If|if-|Never|never-|Time|time-)?(?<name>[^'"`]*)\k<quote>[^)]*\)/g
export const LoaderPlugin = (options: LoaderOptions) => createUnplugin(() => {
  const exclude = options.transform?.exclude || []
  const include = options.transform?.include || []
  const nuxt = tryUseNuxt()

  return {
    name: 'nuxt:components-loader',
    enforce: 'post',
    transformInclude (id) {
      if (exclude.some(pattern => pattern.test(id))) {
        return false
      }
      if (include.some(pattern => pattern.test(id))) {
        return true
      }
      return isVue(id, { type: ['template', 'script'] }) || !!id.match(SX_RE)
    },
    transform (code, id) {
      const components = options.getComponents()

      let num = 0
      const imports = new Set<string>()
      const map = new Map<Component, string>()
      const s = new MagicString(code)
      // replace `_resolveComponent("...")` to direct import
      s.replace(REPLACE_COMPONENT_TO_DIRECT_IMPORT_RE, (full: string, ...args) => {
        const { lazy, modifier, name } = args.pop()
        const normalComponent = findComponent(components, name, options.mode)
        const modifierComponent = !normalComponent && modifier ? findComponent(components, modifier + name, options.mode) : null
        const component = normalComponent || modifierComponent

        if (component) {
          // TODO: refactor to nuxi
          const internalInstall = ((component as any)._internal_install) as string
          if (internalInstall && nuxt?.options.test === false) {
            if (!nuxt.options.dev) {
              const relativePath = relative(nuxt.options.rootDir, id)
              throw new Error(`[nuxt] \`~/${relativePath}\` is using \`${component.pascalName}\` which requires \`${internalInstall}\``)
            }
            installNuxtModule(internalInstall)
          }
          let identifier = map.get(component) || `__nuxt_component_${num++}`
          map.set(component, identifier)

          const isServerOnly = !component._raw && component.mode === 'server' &&
            !components.some(c => c.pascalName === component.pascalName && c.mode === 'client')
          if (isServerOnly) {
            imports.add(genImport(options.serverComponentRuntime, [{ name: 'createServerComponent' }]))
            imports.add(`const ${identifier} = createServerComponent(${JSON.stringify(component.pascalName)})`)
            if (!options.experimentalComponentIslands) {
              logger.warn(`Standalone server components (\`${name}\`) are not yet supported without enabling \`experimental.componentIslands\`.`)
            }
            return identifier
          }

          const isClientOnly = !component._raw && component.mode === 'client'
          if (isClientOnly) {
            imports.add(genImport('#app/components/client-only', [{ name: 'createClientOnly' }]))
            identifier += '_client'
          }

          if (lazy) {
            const dynamicImport = `${genDynamicImport(component.filePath, { interopDefault: false })}.then(c => c.${component.export ?? 'default'} || c)`
            if (modifier && normalComponent) {
              const relativePath = relative(options.srcDir, component.filePath)
              switch (modifier) {
                case 'Visible':
                case 'visible-':
                  imports.add(genImport(options.clientDelayedComponentRuntime, [{ name: 'createLazyVisibleComponent' }]))
                  identifier += '_lazy_visible'
                  imports.add(`const ${identifier} = createLazyVisibleComponent(${JSON.stringify(relativePath)}, ${dynamicImport})`)
                  break
                case 'Interaction':
                case 'interaction-':
                  imports.add(genImport(options.clientDelayedComponentRuntime, [{ name: 'createLazyInteractionComponent' }]))
                  identifier += '_lazy_event'
                  imports.add(`const ${identifier} = createLazyInteractionComponent(${JSON.stringify(relativePath)}, ${dynamicImport})`)
                  break
                case 'Idle':
                case 'idle-':
                  imports.add(genImport(options.clientDelayedComponentRuntime, [{ name: 'createLazyIdleComponent' }]))
                  identifier += '_lazy_idle'
                  imports.add(`const ${identifier} = createLazyIdleComponent(${JSON.stringify(relativePath)}, ${dynamicImport})`)
                  break
                case 'MediaQuery':
                case 'media-query-':
                  imports.add(genImport(options.clientDelayedComponentRuntime, [{ name: 'createLazyMediaQueryComponent' }]))
                  identifier += '_lazy_media'
                  imports.add(`const ${identifier} = createLazyMediaQueryComponent(${JSON.stringify(relativePath)}, ${dynamicImport})`)
                  break
                case 'If':
                case 'if-':
                  imports.add(genImport(options.clientDelayedComponentRuntime, [{ name: 'createLazyIfComponent' }]))
                  identifier += '_lazy_if'
                  imports.add(`const ${identifier} = createLazyIfComponent(${JSON.stringify(relativePath)}, ${dynamicImport})`)
                  break
                case 'Never':
                case 'never-':
                  imports.add(genImport(options.clientDelayedComponentRuntime, [{ name: 'createLazyNeverComponent' }]))
                  identifier += '_lazy_never'
                  imports.add(`const ${identifier} = createLazyNeverComponent(${JSON.stringify(relativePath)}, ${dynamicImport})`)
                  break
                case 'Time':
                case 'time-':
                  imports.add(genImport(options.clientDelayedComponentRuntime, [{ name: 'createLazyTimeComponent' }]))
                  identifier += '_lazy_time'
                  imports.add(`const ${identifier} = createLazyTimeComponent(${JSON.stringify(relativePath)}, ${dynamicImport})`)
                  break
              }
            } else {
              imports.add(genImport('vue', [{ name: 'defineAsyncComponent', as: '__defineAsyncComponent' }]))
              identifier += '_lazy'
              imports.add(`const ${identifier} = __defineAsyncComponent(${dynamicImport}${isClientOnly ? '.then(c => createClientOnly(c))' : ''})`)
            }
          } else {
            imports.add(genImport(component.filePath, [{ name: component._raw ? 'default' : component.export, as: identifier }]))

            if (isClientOnly) {
              imports.add(`const ${identifier}_wrapped = createClientOnly(${identifier})`)
              identifier += '_wrapped'
            }
          }

          return identifier
        }
        // no matched
        return full
      })

      if (imports.size) {
        s.prepend([...imports, ''].join('\n'))
      }

      if (s.hasChanged()) {
        return {
          code: s.toString(),
          map: options.sourcemap
            ? s.generateMap({ hires: true })
            : undefined,
        }
      }
    },
  }
})

function findComponent (components: Component[], name: string, mode: LoaderOptions['mode']) {
  const id = pascalCase(name).replace(QUOTE_RE, '')
  // Prefer exact match
  const component = components.find(component => id === component.pascalName && ['all', mode, undefined].includes(component.mode))
  if (component) { return component }

  const otherModeComponent = components.find(component => id === component.pascalName)

  // Render client-only components on the server with <ServerPlaceholder> (a simple div)
  if (mode === 'server' && otherModeComponent) {
    return components.find(c => c.pascalName === 'ServerPlaceholder')
  }

  // Return the other-mode component in all other cases - we'll handle createClientOnly
  // and createServerComponent above
  return otherModeComponent
}
