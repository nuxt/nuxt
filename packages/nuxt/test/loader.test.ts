import { beforeEach, describe, expect, it, vi } from 'vitest'
import type { Plugin } from 'vite'
import type { Component } from '@nuxt/schema'
import { tryUseNuxt } from '@nuxt/kit'
import { LoaderPlugin } from '../src/components/plugins/loader'
import { normalizeLineEndings } from './utils'
import { join } from 'node:path'
import { distDir } from 'vitest/node'

vi.mock('@nuxt/kit', async (og) => {
  return {
    ...(await og<typeof import('@nuxt/kit')>()),
    tryUseNuxt: vi.fn(),
  }
})

// linux - windows compatibility
vi.mock('node:url', async (og) => {
  return {
    ...(await og<typeof import('node:url')>()),
    pathToFileURL: vi.fn((path: string) => ({ href: path })),
  }
})

const getComponents = () => [{
  filePath: '/root/serverComponent.server.vue',
  mode: 'server',
  pascalName: 'ServerComponent',
  island: true,
  kebabName: 'server-component',
  chunkName: 'components/server-component',
  export: 'default',
  shortPath: '',
  prefetch: false,
  preload: false,
}, {
  filePath: '/root/hello.vue',
  mode: 'server',
  pascalName: 'HelloWorld',
  island: true,
  kebabName: 'hello-world',
  chunkName: 'components/hello',
  export: 'default',
  shortPath: '',
  prefetch: false,
  preload: false,
}, {
  filePath: '/root/Counter.vue',
  mode: 'all',
  pascalName: 'Counter',
  island: false,
  kebabName: 'counter',
  chunkName: 'components/counter',
  export: 'default',
  shortPath: '',
  prefetch: false,
  preload: false,
}] as Component[]

const viteTransform = async (source: string, id: string) => {
  const vitePlugin = LoaderPlugin({
    getComponents,
    mode: 'server',
    serverComponentRuntime: join(distDir, 'components/runtime/server-component'),
    clientDelayedComponentRuntime: join(distDir, 'components/runtime/lazy-hydrated-component'),
    srcDir: '/src'
  }).raw({}, { framework: 'vite' }) as Plugin

  // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
  const result = await (vitePlugin.transform! as Function)(source, id)
  return typeof result === 'string' ? result : result?.code
}

const renderFn = `
function _sfc_ssrRender(_ctx, _push, _parent, _attrs, $props, $setup, $data, $options) {
  const _component_Counter = _resolveComponent("Counter");
  _push(\`<div\`);
  _push(\`</div>\`);
}
`

describe.each([
  { mode: 'server' },
])(`LoaderPlugin`, ({ mode }) => {
  if (mode === 'server') {
    describe('wrap WithIslandTeleport()', () => {
      describe('vite', () => {
        describe('componentIslands === true', () => {
          beforeEach(() => {
            vi.mocked(tryUseNuxt).mockReturnValue({
              options: {
                // @ts-expect-error partial mock
                experimental: {
                  componentIslands: true,
                },
                dev: false,
                test: false,
              },
            })
          })
          it('should not wrap components with WithIslandTeleport', async () => {
            const result = await viteTransform(renderFn, '/root/serverComponent.server.vue')

            expect(normalizeLineEndings(result)).toMatchSnapshot()
            expect(result).not.toContain('withIslandTeleport(__nuxt_component_0)')
          })
        })
        describe('componentIslands.selectiveClient === true', () => {
          beforeEach(() => {
            vi.mocked(tryUseNuxt).mockReturnValue({
              options: {
                // @ts-expect-error partial mock
                experimental: {
                  componentIslands: {
                    selectiveClient: true,
                  },
                },
                dev: false,
                test: false,
              },
            })
          })

          it('should wrap components with WithIslandTeleport in island', async () => {
            const result = await viteTransform(renderFn, '/root/serverComponent.server.vue')

            expect(normalizeLineEndings(result)).toMatchSnapshot()
            expect(result).toContain('withIslandTeleport(__nuxt_component_0)')
          })

          it('should not wrap components with WithIslandTeleport if not in island', async () => {
            const result = await viteTransform(renderFn, 'hello.vue')

            expect(normalizeLineEndings(result)).toMatchSnapshot()
            expect(result).not.toContain('withIslandTeleport(__nuxt_component_0)')
          })
        })
        describe('componentIslands.selectiveClient === \'deep\'', () => {
          beforeEach(() => {
            vi.mocked(tryUseNuxt).mockReturnValue({
              options: {
                // @ts-expect-error partial mock
                experimental: {
                  componentIslands: {
                    selectiveClient: 'deep',
                  },
                },
                dev: false,
                test: false,
              },
            })
          })
          it.each([{ filename: '/root/serverComponent.server.vue' }, { filename: '/root/hello.vue' }])('should wrap components with WithIslandTeleport', async ({ filename }) => {
            const result = await viteTransform(renderFn, filename)

            expect(normalizeLineEndings(result)).toMatchSnapshot()
            expect(result).toContain('withIslandTeleport(__nuxt_component_0)')
          })
        })
      })
    })
  }
})
