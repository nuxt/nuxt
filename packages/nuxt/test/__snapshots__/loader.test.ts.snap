// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoaderPlugin > wrap WithIslandTeleport() > vite > componentIslands === true > should not wrap components with WithIslandTeleport 1`] = `
"import { default as __nuxt_component_0 } from "/root/Counter.vue";

function _sfc_ssrRender(_ctx, _push, _parent, _attrs, $props, $setup, $data, $options) {
  const _component_Counter = __nuxt_component_0;
  _push(\`<div\`);
  _push(\`</div>\`);
}
"
`;

exports[`LoaderPlugin > wrap WithIslandTeleport() > vite > componentIslands.selectiveClient === 'deep' > should wrap components with WithIslandTeleport 1`] = `
"import { default as __nuxt_component_0 } from "/root/Counter.vue";
import { withIslandTeleport } from "#app/components/utils";
const __nuxt_component_0_wrapped = withIslandTeleport(__nuxt_component_0)

function _sfc_ssrRender(_ctx, _push, _parent, _attrs, $props, $setup, $data, $options) {
  const _component_Counter = __nuxt_component_0_wrapped;
  _push(\`<div\`);
  _push(\`</div>\`);
}
"
`;

exports[`LoaderPlugin > wrap WithIslandTeleport() > vite > componentIslands.selectiveClient === 'deep' > should wrap components with WithIslandTeleport 2`] = `
"import { default as __nuxt_component_0 } from "/root/Counter.vue";
import { withIslandTeleport } from "#app/components/utils";
const __nuxt_component_0_wrapped = withIslandTeleport(__nuxt_component_0)

function _sfc_ssrRender(_ctx, _push, _parent, _attrs, $props, $setup, $data, $options) {
  const _component_Counter = __nuxt_component_0_wrapped;
  _push(\`<div\`);
  _push(\`</div>\`);
}
"
`;

exports[`LoaderPlugin > wrap WithIslandTeleport() > vite > componentIslands.selectiveClient === true > should not wrap components with WithIslandTeleport if not in island 1`] = `
"import { default as __nuxt_component_0 } from "/root/Counter.vue";

function _sfc_ssrRender(_ctx, _push, _parent, _attrs, $props, $setup, $data, $options) {
  const _component_Counter = __nuxt_component_0;
  _push(\`<div\`);
  _push(\`</div>\`);
}
"
`;

exports[`LoaderPlugin > wrap WithIslandTeleport() > vite > componentIslands.selectiveClient === true > should wrap components with WithIslandTeleport in island 1`] = `
"import { default as __nuxt_component_0 } from "/root/Counter.vue";
import { withIslandTeleport } from "#app/components/utils";
const __nuxt_component_0_wrapped = withIslandTeleport(__nuxt_component_0)

function _sfc_ssrRender(_ctx, _push, _parent, _attrs, $props, $setup, $data, $options) {
  const _component_Counter = __nuxt_component_0_wrapped;
  _push(\`<div\`);
  _push(\`</div>\`);
}
"
`;
